libevmasm/Assembly.cpp

```diff
@@ -867,175 +954,364 @@ LinkerObject const& Assembly::assemble() const
 	bool setsImmutables = false;
 	bool pushesImmutables = false;
 
-	for (auto const& i: m_items)
-		if (i.type() == AssignImmutable)
-		{
-			i.setImmutableOccurrences(immutableReferencesBySub[i.data()].second.size());
-			setsImmutables = true;
-		}
-		else if (i.type() == PushImmutable)
-			pushesImmutables = true;
+	for (auto const& codeSection: m_codeSections)
+		for (auto const& i: codeSection.items)
+			if (i.type() == AssignImmutable)
+			{
+				i.setImmutableOccurrences(immutableReferencesBySub[i.data()].second.size());
+				setsImmutables = true;
+			}
+			else if (i.type() == PushImmutable)
+				pushesImmutables = true;
 	if (setsImmutables || pushesImmutables)
 		assertThrow(
 			setsImmutables != pushesImmutables,
 			AssemblyException,
 			"Cannot push and assign immutables in the same assembly subroutine."
 		);
 
+	assertThrow(!m_codeSections.empty(), AssemblyException, "Expected at least one code section.");
+	assertThrow(eof || m_codeSections.size() == 1, AssemblyException, "Expected exactly one code section in non-EOF code.");
```

Comment:
```suggestion
	assertThrow(!eof && m_codeSections.size() == 1, AssemblyException, "Expected exactly one code section in non-EOF code.");
```

Not a big deal, but I feel this is more explicit.

Reply:
That's not the same and your version will (obviously) fail for non-eof code :-). This function is shared between legacy and eof - the assertion is to make sure that legacy code only has one code section (i.e. either it's eof code, then fine - or otherwise, i.e. it's legacy code, it has to be only one code section).

Reply:
BTW. I would refactor this function and separated common and uncommon fragments for EOF and legacy, but it can be done in additional PR.

Reply:
How about moving EOF assembly to separated function and leave the legacy as it is? Would it be easier to review?